README
mojolicious-plugin-linkedcontent

This Mojolicious plugin provides five helpers: 

        require_reg
		require_js
        require_css
        include_js
        include_css

Which allows you to manage poading external stuff to page by demand.

Main idea is to load only required external content into html pages.
For example, you have following template system:

        @@ layouts/default.html.ep
        <!doctype html>
        <html>
        <head>
        <%== include_css 'main.css' %>
        </head>
        <body><%= content %></body>
        </html>

        @@ index.html.ep
        % layout 'default';
        % require_css 'ui.css', 'custom-that-requires-ui.css';
        Yea baby!

        @@ no-externals.html.ep
        % layout 'default';
        Clean page!

After render 'index.html.ep' you will see:

        <!doctype html>
        <html>
        <head>
        <link rel='stylesheet' type='text/css' media='screen' href='/css/main.css' />
        <link rel='stylesheet' type='text/css' media='screen' href='/css/ui.css' />
        <link rel='stylesheet' type='text/css' media='screen' href='/css/custom-that-requires-ui.css' />

        </head>
        <body>Yea baby!
        </body>
        </html>

But 'no-externals.html.ep' will give you only:

        <!doctype html>
        <html>
        <head>
        <link rel='stylesheet' type='text/css' media='screen' href='/css/main.css' />

        </head>
        <body>Clean page!
        </body>
        </html>

Using LinkedContent plugin you can easy define default set of css/js resources and append required files
from your templates (even included), actions, plugins, whatever else.

Each resource will be included only once.

You can also create a configuration file like this

my $cdn = '//cdnjs.cloudflare.com/ajax/libs';
{ linkedcontent => 
{
    jquery =>       {
                        js => ["$cdn/jquery/2.2.4/jquery.min"]
                    },
    bootstrap =>    {   
                        js => ["$cdn/twitter-bootstrap/3.3.6/js/bootstrap.min"],
                        css => [
                            "$cdn/twitter-bootstrap/3.3.6/css/bootstrap.min",
                            "$cdn/twitter-bootstrap/3.3.6/css/bootstrap-theme.min",

                        ],
                        deps => ['jquery']
                    },
	fullcalendar => {
                        js => [
                            "$cdn/fullcalendar/2.9.0/fullcalendar.min",
                            "$cdn/fullcalendar/2.9.0/lang/it",
                        ],
                        css => [
                            "$cdn/fullcalendar/2.9.0/fullcalendar",
                            { 
                                href => "$cdn/fullcalendar/2.9.0/fullcalendar.print",
                                media => "print"
                            }
                        ],      
                        deps => ['jquery']
                    },
}

and use 

% require_reg 'bootstrap';

to load, in one shoot, bootstrap javascript and css files and also jquery files automatically.

You can create your own conf files with your preferred CDN, libraries and dependences.

INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

COPYRIGHT AND LICENCE

Copyright (C) 2010 - 2011, Yaroslav Korshak

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.


Changes
Revision history for mojolicious-plugin-linkedcontent

0.08	Mon Jul 11 15:30:00 2016
		Support for registered js and css with dependences as require.js (Emiliano Bruni)
		Extensions .js and .css can be omitted
		Support for css for different media using { href => '...', media => '....' }

0.07    Fri Jan 15 15:56:32 2015
        Support Mojolicious 5.0+
        Replaced render with render_to_string

0.06    Wed Aug 14 22:37:01 2013
        Support Mojolicious 4.0+
        Replaced render_partial with render (oliverguenther)

0.05    Mon Oct 22 23:50:27 2012
        Fixed undocumented stuff broken by Mojolicious 2.62+
        Updated to Module::Install 1.06
        Separated author tests

0.04    Tue Nov 15 11:27:26 2011
        Switched to Module::Install

0.03_1  Thu Feb 10 14:52:06 2011
        Upgraded to Mojolicious 1.01
        Removed Module::Build suport

0.02_1  Fri Dec 10 21:21:52 2010
        CSS support
        Elements ordering fix
        Support for absolute url
        Support for absolute path

0.01_1  Wed Jan 27 18:07:19 2010
        Initial release.


lib/Mojolicious/Plugin/LinkedContent.pm
# Copyright (C) 2010, Yaroslav Korshak - 2021 Emiliano Bruni

package Mojolicious::Plugin::LinkedContent;

use warnings;
use strict;
require Mojo::URL;
use Mojolicious::Plugin::Config;
use LWP::Simple;
use File::Temp;

use base 'Mojolicious::Plugin';

our $VERSION = '0.08';

my %defaults = (
    'js_base'  => '/js',
    'css_base' => '/css',
	'reg_config' => 'https://raw.githubusercontent.com/EmilianoBruni/MPLConfig/main/linked_content.cfg',
);

our $reverse = 0;

my $stashkey = '$linked_store';

sub register {
    my ($self, $app, $params) = @_;
    for (qw/js_base css_base reg_config/) {
        $self->{$_} =
          defined($params->{$_}) ? delete($params->{$_}) : $defaults{$_};
    }

	$self->loaded_reg_items($app);

    push @{$app->renderer->classes}, __PACKAGE__;

    $app->renderer->add_helper(
        require_js => sub {
            $self->store_items('js', @_);
        }
    );
    $app->renderer->add_helper(
        require_css => sub {
            $self->store_items('css', @_);
        }
    );
    $app->renderer->add_helper(
        require_reg => sub {
            $self->store_items_reg(@_);
        }
    );
    $app->renderer->add_helper(
        include_css => sub {
            $self->include_css(@_);
        }
    );
    $app->renderer->add_helper(
        include_js => sub {
            $self->include_js(@_);
        }
    );

    $app->log->debug("Plugin " . __PACKAGE__ . " registred!");
}

sub loaded_reg_items {
	my $s	= shift;
	my $app	= shift;
	$s->{reg_items} = {};
	return unless ($s->{reg_config});

    my $file = $s->{reg_config};

    my $tmp;
    if ($s->_is_remote($file)) {
        # download
        my $content = get($file);
        $tmp = File::Temp->new(SUFFIX => '.cfg' );
        print $tmp $content;
        $file = $tmp->filename;
        close($tmp);
    }

	my $cfg = new Mojolicious::Plugin::Config->new->load($file);

	$s->{reg_items} = $cfg->{linkedcontent} if (exists $cfg->{linkedcontent});
	$app->log->debug("Registry library loaded at " . $s->{reg_config});
}

sub store_items_reg {
    my ($s, $c, @items) = @_;
	foreach my $item (@items) {
		if (exists $s->{reg_items}->{$item}) {
			my $item_info = $s->{reg_items}->{$item};
			if (exists $item_info->{deps}) {
				$s->store_items_reg($c,@{$item_info->{deps}});
			}
			foreach (qw/js css/) {
				$s->store_items($_,$c,@{$item_info->{$_}})
					if exists $item_info->{$_};
			}
		}
	}
}

sub store_items {
    my ($self, $target, $c, @items) = @_;

    my $upd;
    my $store = $c->stash($stashkey) || {};
    for ($reverse ? reverse(@items) : @items) {
        if (exists $store->{'garage'}{$target}{$_}) {
            next unless $reverse;
            my $x = $_;
            @{$store->{'box'}{$target}} = grep $_ ne $x,
              @{$store->{'box'}{$target}};
        }
        $store->{'garage'}{$target}{$_} = 1;
        if (!$reverse) { push(@{$store->{'box'}{$target}}, $_) }
        else           { unshift(@{$store->{'box'}{$target}}, $_); }
    }
    $c->stash($stashkey => $store);
}

sub include_js {
    my $self = shift;
    my $c    = shift;
    local $reverse = 1;
    $self->store_items('js', $c, @_) if @_;
    my $store = $c->stash($stashkey);
    return '' unless $store->{'box'}{'js'};
    my @ct;
    for (@{$store->{'box'}{'js'}}) {

		$_ .= '.js' unless (/\.js$/);

        $c->stash('$linked_item' => $self->_prepend_path($_, 'js_base'));

        push @ct, $c->render_to_string(
            template => 'LinkedContent/js',
            format   => 'html',
            handler  => 'ep',

            # template_class is deprecated since Mojolicious 2.62
            # was removed at some point which broke my code.
            # But it'll live here for a while
            template_class => __PACKAGE__
        );
    }
    $c->stash('$linked_item', undef);
    return join '', @ct;
}

sub include_css {
    my $self = shift;
    my $c    = shift;
    local $reverse = 1;
    $self->store_items('css', $c, @_) if @_;
    my $store = $c->stash($stashkey);
    return '' unless $store->{'box'}{'css'};
    my @ct;
    for (@{$store->{'box'}{'css'}}) {

		my $stash = {
            '$linked_media' => 'screen'
        };

        if (ref($_) eq 'HASH') {
            $stash->{'$linked_item'} = $_->{href};
            $stash->{'$linked_media'} = $_->{media} if (exists $_->{media});
        } else {
            $stash->{'$linked_item'} = $_;
        }

        $stash->{'$linked_item'}  .= '.css' unless ($stash->{'$linked_item'} =~/\.css$/);
        $stash->{'$linked_item'} = $self->_prepend_path($stash->{'$linked_item'}, 'css_base');

        $c->stash($stash);

        push @ct, $c->render_to_string(
            template => 'LinkedContent/css',
            format   => 'html',
            handler  => 'ep',

            # template_class is deprecated since Mojolicious 2.62
            # was removed at some point which broke my code.
            # But it'll live here for a while
            template_class => __PACKAGE__
        );
    }
    $c->stash('$linked_item', undef);
    return join '', @ct;
}

sub _prepend_path {
    my ($self, $path, $base) = @_;

    my $url = Mojo::URL->new($path);
    if ($url->is_abs || $url->path->leading_slash) {

        # Absolute path or absolute url returned as is
        return $path;
    }

    # Basepath not defined
    return unless $self->{$base};

    # Prepend path with base
    my $basepath = Mojo::Path->new($self->{$base});
    unshift @{$url->path->parts}, @{$basepath->parts};

    # Inherit leading slash from basepath
    $url->path->leading_slash($basepath->leading_slash);

    return $url->to_string;
}

sub _is_remote  {
    my ($s, $file) = (shift, shift);

    return 1 if ($file =~ /^http/);
}
1;

__DATA__
@@ LinkedContent/js.html.ep
<script src='<%== $self->stash('$linked_item') %>'></script>
@@ LinkedContent/css.html.ep
<link rel='stylesheet' type='text/css' media='<%= $self->stash('$linked_media') %>' href='<%= $self->stash('$linked_item') %>' />
__END__

=encoding utf8


=head1 NAME

Mojolicious::Plugin::LinkedContent - manage linked css and js


=head1 SYNOPSIS

    use base 'Mojolicious';
    sub statup {
        my $self = shift;
        $self->plugin( 'linked_content' );
    }

Somewhere in template:

    % require_css 'mypage.css';
    % require_js 'myscript.js';
    % require_reg 'bootstrap';

And in <HEAD> of your layout:

    %== include_css;
    %== include_js;


=head1 DESCRIPTION

Helps to manage scripts and styles included in document.

=head1 INTERFACE

=head1 HELPERS

=over

=item require_js

Add one or more js files to load queue.

=item require_css

Add one or more css files to load queue.

=item require_reg

Add a library and its dependences based on reg_config file

=item register

Render the plugin.
Internal

=item include_js
=item include_css

Render queue to template

=back

=head2 ITEMS

=over

=item store_items

Internal method

=back

=head1 CONFIGURATION AND ENVIRONMENT

L<Mojolicious::Plugin::LinkedContent> can recieve parameters
when loaded from  L<Mojolicious> like this:

    $self->plugin(
        'linked_content',
        'js_base'  => '/jsdir',
        'css_base' => '/cssdir'
        'reg_config' => '/linked_content.cfg',
    );

If no basedirs provided, '/js' and '/css' used by default.
If no reg_config is provided a cloud example file is used.
Default reg_config URL: https://raw.githubusercontent.com/EmilianoBruni/MPLConfig/main/linked_content.cfg

=head1 AUTHOR

Yaroslav Korshak  C<< <ykorshak@gmail.com> >>,
Emiliano Bruni C<< <info@ebruni.it >>

=head1 CREDITS

=over 2

Oliver Günther

=back

=head1 LICENCE AND COPYRIGHT

Copyright (C) 2010 - 2013, Yaroslav Korshak
Copyright (C) 2019 - 2021, Emiliano Bruni

This module is free software; you can redistribute it and/or
modify it under the same terms as Perl itself. See L<perlartistic>.
